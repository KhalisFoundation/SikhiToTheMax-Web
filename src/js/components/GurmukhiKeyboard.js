import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';
import ArrowIcon from './Icons/Arrow';

const ButtonList = ({ buttons = [] }) =>
  buttons.map((button, i) => (
    <button key={i} type="button">
      {button}
    </button>
  ));

export default class GurmukhiKeyboard extends React.PureComponent {
  static propTypes = {
    value: PropTypes.string.isRequired,
    active: PropTypes.bool.isRequired,
    onKeyClick: PropTypes.func.isRequired,
    onClose: PropTypes.func.isRequired,
  };

  // Instead of attaching click to each button, we are using event delegation.
  click = ({ target: _$button, currentTarget: $keyboard }) => {
    const { value, onKeyClick, onClose } = this.props;
    const $button =
      _$button.nodeName.toUpperCase() === 'I' ? _$button.parentNode : _$button;
    const { action } = $button.dataset;

    switch (action) {
      case 'close': {
        onClose();
        break;
      }
      case 'bksp': {
        onKeyClick(value.substring(0, value.length - 1));
        break;
      }
      case 'page-1':
      case 'page-2': {
        [...$keyboard.querySelectorAll('.page')].forEach(e => {
          e.style.display = 'none';
        });
        document.getElementById(`gurmukhi-keyboard-${action}`).style.display =
          'block';
        break;
      }
      case 'help':
        break;
      default: {
        onKeyClick(value + ($button.dataset.value || $button.innerText));
      }
    }
  };

  render() {
    return (
      <div
        className={`gurmukhi-keyboard gurbani-font ${
          this.props.active ? 'active' : ''
        }`}
        onClick={this.click}
      >
        <div
          className="page"
          style={{ display: 'block' }}
          id="gurmukhi-keyboard-page-1"
        >
          <div className="keyboard-row">
            <div className="keyboard-row-set">
              <ButtonList buttons={['a', 'A', 'e', 's', 'h']} />
            </div>
            <div className="keyboard-row-set">
              <ButtonList buttons={['k', 'K', 'g', 'G', '|']} />
            </div>
          </div>
          <div className="keyboard-row">
            <div className="keyboard-row-set">
              <ButtonList buttons={['c', 'C', 'j', 'J', '\\']} />
            </div>
            <div className="keyboard-row-set">
              <ButtonList buttons={['t', 'T', 'f', 'F', 'x']} />
            </div>
          </div>
          <div className="keyboard-row">
            <div className="keyboard-row-set">
              <ButtonList buttons={['q', 'Q', 'd', 'D', 'n']} />
            </div>
            <div className="keyboard-row-set">
              <ButtonList buttons={['p', 'P', 'b', 'B', 'm']} />
            </div>
          </div>
          <div className="keyboard-row">
            <div className="keyboard-row-set">
              <ButtonList buttons={['X', 'r', 'l', 'v', 'V']} />
            </div>
            <div className="keyboard-row-set">
              <button type="button">{'\u00a0'}</button>
              <button type="button" data-action="page-1" className="active">
                1
              </button>
              <button type="button" data-action="page-2">
                2
              </button>
              <button type="button" data-action="bksp">
                <ArrowIcon />
              </button>

              <Link to="/help#Web-how-to-type-gurmukhi-with-keyboard">
                <button data-action="help" type="button">
                  ?
                </button>
              </Link>
            </div>
          </div>
        </div>
        <div className="page" id="gurmukhi-keyboard-page-2">
          <div className="keyboard-row">
            <div className="keyboard-row-set">
              <ButtonList buttons={[1, 2, 3, 4, 5]} />
            </div>
            <div className="keyboard-row-set">
              <ButtonList buttons={[6, 7, 8, 9, 0]} />
            </div>
          </div>
          <div className="keyboard-row">
            <div className="keyboard-row-set">
              <ButtonList buttons={['w', 'i', 'I', 'u', 'U']} />
            </div>
            <div className="keyboard-row-set">
              <ButtonList buttons={['y', 'Y', 'o', 'O', 'M']} />
            </div>
          </div>
          <div className="keyboard-row">
            <div className="keyboard-row-set">
              <ButtonList buttons={['N', 'W', '`', '~', 'R']} />
            </div>
            <div className="keyboard-row-set">
              <ButtonList buttons={['H', '˜', '´', 'Í', 'Ï']} />
            </div>
          </div>
          <div className="keyboard-row">
            <div className="keyboard-row-set">
              <ButtonList buttons={['ç', 'E', '^', '\u00a0', '\u00a0']} />
            </div>
            <div className="keyboard-row-set">
              <button type="button">{'\u00a0'}</button>
              <button type="button" data-action="page-1">
                1
              </button>
              <button type="button" data-action="page-2" className="active">
                2
              </button>
              <button type="button" data-action="bksp">
                <ArrowIcon />
              </button>
              <Link to="/help#Web-how-to-type-gurmukhi-with-keyboard">
                <button data-action="help" type="button">
                  ?
                </button>
              </Link>
            </div>
          </div>
        </div>
      </div>
    );
  }
}
