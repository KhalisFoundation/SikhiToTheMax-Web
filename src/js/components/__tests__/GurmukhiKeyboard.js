/* global describe, it, expect, jest */
import React from 'react';
import { mount } from 'enzyme';
import GurmukhiKeyboard from '../GurmukhiKeyboard';
import { StaticRouter } from 'react-router-dom';

describe('<GurmukhiKeyboard />', () => {
  it('renders correctly', () => {
    const component = mount(
      <StaticRouter>
        <GurmukhiKeyboard value={'mere'} onKeyClick={jest.fn()} />
      </StaticRouter>
    );

    expect(component).toMatchSnapshot();
  });
  it('triggers onKeyClick', () => {
    const onKeyClick = jest.fn();
    const component = mount(
      <StaticRouter>
        <GurmukhiKeyboard onKeyClick={onKeyClick} value={'mere'} />
      </StaticRouter>
    );

    const button = component.find('button').filterWhere(n => n.text() === 'A');

    button.simulate('click');

    expect(onKeyClick).toHaveBeenCalledWith('mereA');

    component
      .find('[title="Backspace"]')
      .first()
      .simulate('click');

    expect(onKeyClick).toHaveBeenCalledWith('mer');
  });
});
