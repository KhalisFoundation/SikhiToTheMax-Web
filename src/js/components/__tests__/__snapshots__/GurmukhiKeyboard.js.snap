// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GurmukhiKeyboard /> renders correctly 1`] = `
<StaticRouter
  basename=""
  context={Object {}}
  location="/"
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <GurmukhiKeyboard
      active={true}
      onKeyClick={[MockFunction]}
      value="mere"
    >
      <div
        className="gurmukhi-keyboard gurbani-font active"
      >
        <div
          className="page"
          id="gurmukhi-keyboard-page-1"
          key="0"
        >
          <div
            className="keyboard-row"
            key="0-0"
          >
            <div
              className="keyboard-row-set"
            >
              <ButtonList
                buttons={
                  Array [
                    "a",
                    "A",
                    "e",
                    "s",
                    "h",
                  ]
                }
                onButtonClick={[Function]}
              >
                <button
                  data-value="a"
                  key="0"
                  onClick={[Function]}
                  type="button"
                >
                  a
                </button>
                <button
                  data-value="A"
                  key="1"
                  onClick={[Function]}
                  type="button"
                >
                  A
                </button>
                <button
                  data-value="e"
                  key="2"
                  onClick={[Function]}
                  type="button"
                >
                  e
                </button>
                <button
                  data-value="s"
                  key="3"
                  onClick={[Function]}
                  type="button"
                >
                  s
                </button>
                <button
                  data-value="h"
                  key="4"
                  onClick={[Function]}
                  type="button"
                >
                  h
                </button>
              </ButtonList>
            </div>
            <div
              className="keyboard-row-set"
            >
              <ButtonList
                buttons={
                  Array [
                    "k",
                    "K",
                    "g",
                    "G",
                    "|",
                  ]
                }
                onButtonClick={[Function]}
              >
                <button
                  data-value="k"
                  key="0"
                  onClick={[Function]}
                  type="button"
                >
                  k
                </button>
                <button
                  data-value="K"
                  key="1"
                  onClick={[Function]}
                  type="button"
                >
                  K
                </button>
                <button
                  data-value="g"
                  key="2"
                  onClick={[Function]}
                  type="button"
                >
                  g
                </button>
                <button
                  data-value="G"
                  key="3"
                  onClick={[Function]}
                  type="button"
                >
                  G
                </button>
                <button
                  data-value="|"
                  key="4"
                  onClick={[Function]}
                  type="button"
                >
                  |
                </button>
              </ButtonList>
            </div>
          </div>
          <div
            className="keyboard-row"
            key="0-1"
          >
            <div
              className="keyboard-row-set"
            >
              <ButtonList
                buttons={
                  Array [
                    "c",
                    "C",
                    "j",
                    "J",
                    "\\\\",
                  ]
                }
                onButtonClick={[Function]}
              >
                <button
                  data-value="c"
                  key="0"
                  onClick={[Function]}
                  type="button"
                >
                  c
                </button>
                <button
                  data-value="C"
                  key="1"
                  onClick={[Function]}
                  type="button"
                >
                  C
                </button>
                <button
                  data-value="j"
                  key="2"
                  onClick={[Function]}
                  type="button"
                >
                  j
                </button>
                <button
                  data-value="J"
                  key="3"
                  onClick={[Function]}
                  type="button"
                >
                  J
                </button>
                <button
                  data-value="\\\\"
                  key="4"
                  onClick={[Function]}
                  type="button"
                >
                  \\
                </button>
              </ButtonList>
            </div>
            <div
              className="keyboard-row-set"
            >
              <ButtonList
                buttons={
                  Array [
                    "t",
                    "T",
                    "f",
                    "F",
                    "x",
                  ]
                }
                onButtonClick={[Function]}
              >
                <button
                  data-value="t"
                  key="0"
                  onClick={[Function]}
                  type="button"
                >
                  t
                </button>
                <button
                  data-value="T"
                  key="1"
                  onClick={[Function]}
                  type="button"
                >
                  T
                </button>
                <button
                  data-value="f"
                  key="2"
                  onClick={[Function]}
                  type="button"
                >
                  f
                </button>
                <button
                  data-value="F"
                  key="3"
                  onClick={[Function]}
                  type="button"
                >
                  F
                </button>
                <button
                  data-value="x"
                  key="4"
                  onClick={[Function]}
                  type="button"
                >
                  x
                </button>
              </ButtonList>
            </div>
          </div>
          <div
            className="keyboard-row"
            key="0-2"
          >
            <div
              className="keyboard-row-set"
            >
              <ButtonList
                buttons={
                  Array [
                    "q",
                    "Q",
                    "d",
                    "D",
                    "n",
                  ]
                }
                onButtonClick={[Function]}
              >
                <button
                  data-value="q"
                  key="0"
                  onClick={[Function]}
                  type="button"
                >
                  q
                </button>
                <button
                  data-value="Q"
                  key="1"
                  onClick={[Function]}
                  type="button"
                >
                  Q
                </button>
                <button
                  data-value="d"
                  key="2"
                  onClick={[Function]}
                  type="button"
                >
                  d
                </button>
                <button
                  data-value="D"
                  key="3"
                  onClick={[Function]}
                  type="button"
                >
                  D
                </button>
                <button
                  data-value="n"
                  key="4"
                  onClick={[Function]}
                  type="button"
                >
                  n
                </button>
              </ButtonList>
            </div>
            <div
              className="keyboard-row-set"
            >
              <ButtonList
                buttons={
                  Array [
                    "p",
                    "P",
                    "b",
                    "B",
                    "m",
                  ]
                }
                onButtonClick={[Function]}
              >
                <button
                  data-value="p"
                  key="0"
                  onClick={[Function]}
                  type="button"
                >
                  p
                </button>
                <button
                  data-value="P"
                  key="1"
                  onClick={[Function]}
                  type="button"
                >
                  P
                </button>
                <button
                  data-value="b"
                  key="2"
                  onClick={[Function]}
                  type="button"
                >
                  b
                </button>
                <button
                  data-value="B"
                  key="3"
                  onClick={[Function]}
                  type="button"
                >
                  B
                </button>
                <button
                  data-value="m"
                  key="4"
                  onClick={[Function]}
                  type="button"
                >
                  m
                </button>
              </ButtonList>
            </div>
          </div>
          <div
            className="keyboard-row"
            key="0-3"
          >
            <div
              className="keyboard-row-set"
            >
              <ButtonList
                buttons={
                  Array [
                    "X",
                    "r",
                    "l",
                    "v",
                    "V",
                  ]
                }
                onButtonClick={[Function]}
              >
                <button
                  data-value="X"
                  key="0"
                  onClick={[Function]}
                  type="button"
                >
                  X
                </button>
                <button
                  data-value="r"
                  key="1"
                  onClick={[Function]}
                  type="button"
                >
                  r
                </button>
                <button
                  data-value="l"
                  key="2"
                  onClick={[Function]}
                  type="button"
                >
                  l
                </button>
                <button
                  data-value="v"
                  key="3"
                  onClick={[Function]}
                  type="button"
                >
                  v
                </button>
                <button
                  data-value="V"
                  key="4"
                  onClick={[Function]}
                  type="button"
                >
                  V
                </button>
              </ButtonList>
            </div>
            <div
              className="keyboard-row-set"
            >
              <button
                type="button"
              >
                 
              </button>
              <button
                className="active"
                data-action="page-1"
                key="1"
                onClick={[Function]}
                title="Page 1"
                type="button"
              >
                1
              </button>
              <button
                className=""
                data-action="page-2"
                key="2"
                onClick={[Function]}
                title="Page 2"
                type="button"
              >
                2
              </button>
              <button
                data-action="bksp"
                onClick={[Function]}
                title="Backspace"
                type="button"
              >
                <Arrow>
                  <svg
                    height="1em"
                    viewBox="0 0 448 512"
                    width="1em"
                  >
                    <path
                      d="M134.059 296H436c6.627 0 12-5.373 12-12v-56c0-6.627-5.373-12-12-12H134.059v-46.059c0-21.382-25.851-32.09-40.971-16.971L7.029 239.029c-9.373 9.373-9.373 24.569 0 33.941l86.059 86.059c15.119 15.119 40.971 4.411 40.971-16.971V296z"
                    />
                  </svg>
                </Arrow>
              </button>
              <Link
                replace={false}
                to="/help#Web-how-to-type-gurmukhi-with-keyboard"
              >
                <a
                  href="/help#Web-how-to-type-gurmukhi-with-keyboard"
                  onClick={[Function]}
                >
                  <button
                    type="button"
                  >
                    ?
                  </button>
                </a>
              </Link>
            </div>
          </div>
        </div>
      </div>
    </GurmukhiKeyboard>
  </Router>
</StaticRouter>
`;
